jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/RabitTracker.java

```diff
@@ -105,123 +45,51 @@ public void uncaughtException(Thread t, Throwable e) {
-        }
-        String[] sep = line.split("=");
-        if (sep.length == 2) {
-          envs.put(sep[0], sep[1]);
-        }
-      }
-      reader.close();
-    } catch (IOException ioe){
-      logger.error("cannot get runtime configuration from tracker process");
-      ioe.printStackTrace();
-      throw ioe;
-    }
-  }
-
-  /** visible for testing */
-  public String getRabitTrackerCommand() {
-    StringBuilder sb = new StringBuilder();
-    if (pythonExec == null || pythonExec.isEmpty()) {
-      sb.append("python ");
-    } else {
-      sb.append(pythonExec + " ");
-    }
-    sb.append(" " + tracker_py + " ");
-    sb.append(" --log-level=DEBUG" + " ");
-    sb.append(" --num-workers=" + numWorkers + " ");
-
-    // we first check the property then check the parameter
-    String hostIpFromProperties = trackerProperties.getHostIp();
-    if(hostIpFromProperties != null && !hostIpFromProperties.isEmpty()) {
-      logger.debug("Using provided host-ip: " + hostIpFromProperties + " from properties");
-      sb.append(" --host-ip=" + hostIpFromProperties + " ");
-    } else if (hostIp != null & !hostIp.isEmpty()) {
-      logger.debug("Using the parametr host-ip: " + hostIp);
-      sb.append(" --host-ip=" + hostIp + " ");
-    }
-    return sb.toString();
-  }
-
-  private boolean startTrackerProcess() {
+  public Map<String, Object> workerArgs() throws XGBoostError {
+    // fixme: timeout
+    String[] args = new String[1];
+    XGBoostJNI.checkCall(XGBoostJNI.TrackerWorkerArgs(this.handle, 0, args));
+    ObjectMapper mapper = new ObjectMapper();
+    TypeReference<Map<String, Object>> typeRef = new TypeReference<Map<String, Object>>() {
+    };
+    Map<String, Object> config;
     try {
-      String cmd = getRabitTrackerCommand();
-      trackerProcess.set(Runtime.getRuntime().exec(cmd));
-      loadEnvs(trackerProcess.get().getInputStream());
-      return true;
-    } catch (IOException ioe) {
-      ioe.printStackTrace();
-      return false;
+      config = mapper.readValue(args[0], typeRef);
+    } catch (JsonProcessingException ex) {
+      throw new XGBoostError("Failed to get worker arguments.", ex);
     }
+    return config;
   }
 
-  public void stop() {
-    if (trackerProcess.get() != null) {
-      trackerProcess.get().destroy();
-    }
+  public void stop() throws XGBoostError {
+    XGBoostJNI.checkCall(XGBoostJNI.TrackerFree(this.handle));
   }
 
-  public boolean start(long timeout) {
-    if (timeout > 0L) {
-      logger.warn("Python RabitTracker does not support timeout. " +
-              "The tracker will wait for all workers to connect indefinitely, unless " +
-              "it is interrupted manually. Use the Scala RabitTracker for timeout support.");
-    }
+  public boolean start() throws XGBoostError {
+    XGBoostJNI.checkCall(XGBoostJNI.TrackerRun(this.handle));
+    this.tracker_daemon = new Thread(() -> {
+      try {
+        XGBoostJNI.checkCall(XGBoostJNI.TrackerWaitFor(this.handle, 0));
```

Comment:
May I ask why tracker_daemon is needed here?

Reply:
I kept it as a handle for the implementation of `uncaughtException`, do you have suggestions for handling it differently?